File: clojure/hamming/src/hamming.clj
(ns hamming)

(defn- same-length?
  [first second]
  (=
    (count first)
    (count second)))

(defn- zip
  [first second]
  (map vector first second))

(defn- same?
  [[a b]]
  (= a b))

(defn- difference
  [first second]
  (->>
    (zip first second)
    (remove same?)
    (count)))

(defn distance
  [first second]
  (if (same-length? first second)
    (difference first second)
    nil))

File: haskell/hamming/src/Hamming.hs
module Hamming (distance) where

distance :: String -> String -> Maybe Int
distance xs ys =
  let
    calculate a b count =
      case (a, b) of
        ([], []) ->
          Just count
        ((a:aTail), (b:bTail)) ->
          calculate aTail bTail (if a == b then count else count + 1)
        _ ->
          Nothing
  in
    calculate xs ys 0
